// redux-saga-requests-axios/lib/axios-driver.js
"use strict";

exports.__esModule = true;
exports.createDriver = void 0;

var _axios = _interopRequireDefault(require("axios"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var prepareSuccessPayload = function prepareSuccessPayload(response) {
  return response.data;
};

var createDriver = function createDriver(axiosInstance) {
  return {
    requestInstance: axiosInstance,
    getAbortSource: function getAbortSource() {
      return _axios["default"].CancelToken.source();
    },
    abortRequest: function abortRequest(abortSource) {
      abortSource.cancel();
    },
    sendRequest: function sendRequest(requestConfig, abortSource) {
      return axiosInstance(_extends({
        cancelToken: abortSource.token
      }, requestConfig));
    },
    getSuccessPayload: function getSuccessPayload(response) {
      if (Array.isArray(response)) {
        return response.map(prepareSuccessPayload);
      }

      return prepareSuccessPayload(response);
    },
    getErrorPayload: function getErrorPayload(error) {
      return error;
    }
  };
};

exports.createDriver = createDriver;
