// redux-saga-requests/es/middleware/requests-cache-middleware.js
import _extends from "@babel/runtime/helpers/extends";
import { GET_REQUEST_CACHE, CLEAR_REQUESTS_CACHE } from '../constants';
import { isRequestAction, isSuccessAction, getRequestActionFromResponse, getActionPayload } from '../actions';

var isCacheValid = function isCacheValid(cache) {
  return cache.expiring === null || Date.now() <= cache.expiring;
};

var getNewCacheTimeout = function getNewCacheTimeout(cache) {
  return cache === true ? null : cache * 1000 + Date.now();
};

var getCacheKey = function getCacheKey(action) {
  return action.type + (action.meta.cacheKey || '');
};

export default (function () {
  var cacheMap = new Map();
  return function () {
    return function (next) {
      return function (action) {
        if (action.type === GET_REQUEST_CACHE) {
          return cacheMap;
        }

        if (action.type === CLEAR_REQUESTS_CACHE) {
          if (action.actionTypes.length === 0) {
            cacheMap.clear();
          } else {
            action.actionTypes.forEach(function (actionType) {
              return cacheMap["delete"](actionType);
            });
          }

          return null;
        }

        if (isRequestAction(action) && action.meta && action.meta.cache) {
          var cacheKey = getCacheKey(action);
          var cacheValue = cacheMap.get(cacheKey);

          if (cacheValue && isCacheValid(cacheValue)) {
            return next(_extends({}, action, {
              meta: _extends({}, action.meta, {
                cacheResponse: cacheValue.response
              })
            }));
          } else if (cacheValue && !isCacheValid(cacheValue)) {
            cacheMap["delete"](cacheKey);
          }
        } else if (isSuccessAction(action) && action.meta && action.meta.cache && !action.meta.cacheResponse) {
          var requestAction = getRequestActionFromResponse(action);

          if (action.meta.cacheKey && action.meta.cacheSize) {
            var currentCacheKeys = Array.from(cacheMap.keys()).filter(function (k) {
              return k.startsWith(requestAction.type);
            });

            if (action.meta.cacheSize === currentCacheKeys.length) {
              cacheMap["delete"](currentCacheKeys[0]);
            }
          }

          cacheMap.set(getCacheKey(requestAction), {
            response: getActionPayload(action).response,
            expiring: getNewCacheTimeout(action.meta.cache)
          });
        }

        return next(action);
      };
    };
  };
});
