// redux-saga-requests/lib/reducers/requests-reducer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _extends4 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _actions = require("../actions");

var _mutationsReducer = _interopRequireDefault(require("./mutations-reducer"));

var _defaultConfig = _interopRequireDefault(require("./default-config"));

// to support libraries like redux-act and redux-actions
var normalizeActionType = function normalizeActionType(actionType) {
  return typeof actionType === 'function' ? actionType.toString() : actionType;
};

var mutationConfigHasRequestKey = function mutationConfigHasRequestKey(config) {
  return typeof config !== 'boolean' && !!config.getRequestKey;
};

var getInitialState = function getInitialState(_ref) {
  var getDefaultData = _ref.getDefaultData,
      multiple = _ref.multiple,
      mutations = _ref.mutations,
      handleMutationsState = _ref.handleMutationsState;
  return {
    data: getDefaultData(multiple),
    pending: 0,
    error: null,
    mutations: handleMutationsState && mutations ? Object.entries(mutations).filter(function (_ref2) {
      var v = _ref2[1];
      return !v.local;
    }).reduce(function (prev, _ref3) {
      var _extends2;

      var k = _ref3[0],
          v = _ref3[1];
      return (0, _extends4["default"])({}, prev, (_extends2 = {}, _extends2[k] = mutationConfigHasRequestKey(v) ? {} : {
        error: null,
        pending: 0
      }, _extends2));
    }, {}) : null
  };
};

var getDataUpdater = function getDataUpdater(reducerConfig, mutationConfig) {
  if (mutationConfig === true || mutationConfig.updateData === true) {
    return reducerConfig.updateData || reducerConfig.getData;
  } else if (typeof mutationConfig === 'function') {
    return mutationConfig;
  } else if (typeof mutationConfig !== 'boolean' && typeof mutationConfig.updateData === 'function') {
    return mutationConfig.updateData;
  }

  return null;
};

var requestMutationReducer = function requestMutationReducer(state, action, config) {
  var mutationConfig = config.mutations[action.type];

  if (mutationConfig.updateDataOptimistic) {
    return (0, _extends4["default"])({}, state, {
      data: mutationConfig.updateDataOptimistic(state, action, config)
    });
  }

  if (mutationConfig.local) {
    var dataUpdater = getDataUpdater(config, mutationConfig);
    return (0, _extends4["default"])({}, state, {
      data: dataUpdater(state, action, config)
    });
  }

  return state;
};

var responseMutationReducer = function responseMutationReducer(state, action, config) {
  var requestAction = (0, _actions.getRequestActionFromResponse)(action);
  var mutationConfig = config.mutations[requestAction.type];

  if ((0, _actions.isSuccessAction)(action)) {
    var dataUpdater = getDataUpdater(config, mutationConfig);
    return dataUpdater ? (0, _extends4["default"])({}, state, {
      data: dataUpdater(state, action, config)
    }) : state;
  } // error or abort case


  return mutationConfig.revertData ? (0, _extends4["default"])({}, state, {
    data: mutationConfig.revertData(state, action, config)
  }) : state;
};

var _default = function _default(localConfig) {
  var config = (0, _extends4["default"])({}, _defaultConfig["default"], {}, localConfig);
  var normalizedActionType = normalizeActionType(config.actionType);
  var shouldActionBeReset = typeof config.resetOn === 'function' ? config.resetOn : function (action) {
    return config.resetOn.map(normalizeActionType).includes(action.type);
  };
  return function (state, action) {
    if (action.meta && action.meta.mutations && normalizedActionType in action.meta.mutations && (!config.mutations || !(normalizedActionType in config.mutations))) {
      var _extends3;

      config.mutations = (0, _extends4["default"])({}, config.mutations, (_extends3 = {}, _extends3[action.type] = action.meta.mutations[normalizedActionType], _extends3));
    }

    var nextState = state || getInitialState(config);

    if (shouldActionBeReset(action)) {
      nextState = (0, _extends4["default"])({}, getInitialState(config), {
        pending: nextState.pending
      });
    }

    var requestAction = (0, _actions.isResponseAction)(action) ? (0, _actions.getRequestActionFromResponse)(action) : action;

    if (config.mutations && requestAction.type in config.mutations) {
      return (0, _extends4["default"])({}, (0, _actions.isResponseAction)(action) ? responseMutationReducer(nextState, action, config) : requestMutationReducer(nextState, action, config), {
        mutations: config.handleMutationsState ? (0, _mutationsReducer["default"])(nextState.mutations, action, config, config.mutations[requestAction.type]) : null
      });
    }

    switch (action.type) {
      case normalizedActionType:
        return config.onRequest(nextState, action, config);

      case (0, _actions.success)(normalizedActionType):
        return config.onSuccess(nextState, action, config);

      case (0, _actions.error)(normalizedActionType):
        return config.onError(nextState, action, config);

      case (0, _actions.abort)(normalizedActionType):
        return config.onAbort(nextState, action, config);

      default:
        return nextState;
    }
  };
};

exports["default"] = _default;
